### PR TITLE
livecheck: support throttle DSL

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Essentially, we need to figure out best way to combine autobump, livecheck, and throttle so that autobumps can properly handle this. Preferably avoiding code duplication across various parts (i.e. bump commands, livecheck logic, github action, etc.)

Some ideas:
1. **Throttle DSL** - This approach
2. Livecheck throttling - ~~e.g. https://github.com/Homebrew/homebrew-core/commit/184767f00391c1aae6dbbfd3a03388c2c5b42857
3. Time-based throttling - would need to change version-based throttling to time-based (days between bump) in JSON, provide a way of detecting some datetime info (e.g. modulo current day of year, attempt during specific timeslot in a day). 
    - Con: This functionality mainly works when on autobump, but is trickier when user bumping is allowed as cannot guarantee cadence here.

### CODE REVIEW

**Review by carlocab — COMMENTED** [review]
Seems sensible to me, but I think we want to deprecate the throttle JSON in favour of this so we don't have two different lists to carry around.

**Comment by cho-m** [comment]
> deprecate the throttle JSON in favour of this so we don't have two different lists to carry around.

That was my plan if we go this route. But may be a follow up so we can migrate `homebrew/core` usage and then deprecate support. I need to figure out how to handle `renovate` and `vim` since we don't collect sufficient version info.

File: Library/Homebrew/dev-cmd/bump.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump.rb:257]: Feels like it'd be nicer to keep this as `T.any(Version, String)` and use a null version or empty string?

File: Library/Homebrew/dev-cmd/bump.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump.rb:366]: Doesn't need cleaned up in this PR but: I hate that it's using different types here to indicate success/failure.

File: Library/Homebrew/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck.rb:99]: Would be nicer to avoid `unsafe` here if possible.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good! There's a bunch of suboptimal type usage in existing code that would be nice to clean up but doesn't need to block this PR.

File: Library/Homebrew/dev-cmd/bump.rb
Comment (cho-m) [reply Library/Homebrew/dev-cmd/bump.rb:257]: Sure. I was originally thinking of passing both versions (unthrottled, throttled) to print output but would require some extra work.

Right now, I am not using both so it makes sense to collapse it back and just return one of them.

**Review by samford — COMMENTED** [review]
From a quick review/test, we still need to add some throttle information to the debug and JSON output. Namely:

* The JSON output is missing an (optional) `latest_throttled` field. The value is present in the data from `#latest_version` but we need to handle it in `#run_checks`.
* We should print the `throttle` value in the debug output.
* We should include the `throttle` value in the `meta` part of the JSON output.

These are simple changes and I worked them up while reviewing this, so I'll just go ahead and push a commit to take care of it (since I can't add suggestions to all the code here).

-----

Past that, this seems reasonable to me. With these changes, livecheck continues to return the latest version and the latest throttled version is simply provided as supplemental information, which is how I would handle this. `brew bump` benefits from the extra information while livecheck continues to return the latest version 👍

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by samford** [comment]
> @samford can you ✅ in future when you feel that way? Thanks 🙇🏻

My wording may have been soft/vague but I still wanted to see the `throttle` variable/method in the `Livecheck` DSL reordered before approving/merging. I pushed a commit to take care of that, so I'll go ahead and approve.

I also rebased this and resolved the merge conflicts, so it should pass CI this time around.

**Review by samford — APPROVED** [review]

File: Library/Homebrew/livecheck.rb
Comment (reitermarkus) [inline Library/Homebrew/livecheck.rb:144]:
```suggestion
      rate: Integer,
```

File: Library/Homebrew/livecheck.rb
Comment (samford) [reply Library/Homebrew/livecheck.rb]: This is ` T.nilable(Integer)` because this method is also used to get the `@throttle` value (i.e., we use `throttle 10` in the DSL and `formula.livecheck.throttle` to obtain the value).

Comment (reitermarkus) [reply Library/Homebrew/livecheck.rb]: That still works if `rate: Integer`, because it is only checked when passed in from the outside.

Comment (reitermarkus) [reply Library/Homebrew/livecheck.rb]: I.e., currently all of these are allowed `throttle 1`, `throttle nil` and `throttle`, by specifying `rate: Integer`, we disallow passing `nil` explicitly.

Comment (samford) [reply Library/Homebrew/livecheck.rb]: Makes sense. Thanks for clarifying. I pushed a commit to incorporate this change.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (cho-m) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:475]: I'll probably flip this (since we want to prioritize DSL now):
```suggestion
        throttled_rate = formula.livecheck.throttle
        throttled_rate ||= formula.tap.audit_exceptions.dig(:throttled_formulae, formula.name)
```

File: Library/Homebrew/dev-cmd/bump.rb
Comment (cho-m) [inline Library/Homebrew/dev-cmd/bump.rb:365]: Given that `brew bump` doesn't really need unthrottled info (and it is available via `livecheck`), I'll go ahead an simplify this (and handle in function)
```suggestion
            livecheck_latest = livecheck_result(loaded_formula_or_cask)
```

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/livecheck.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck.rb]: > Whichever we choose should be consistent with Livecheck code so can discuss in follow up.

Agreed.

Generally I favour (in descending order of preference):
- non-`nilable` types
- `nilable` types
- use of `unsafe`
- use of `must`

**Comment by cho-m** [comment]
Going to have to figure out why test for ZSH completions is failing. I don't think it is related to this PR given we didn't modify `description`. Can't repro locally.
```
Error: Homebrew::Completions when generating completions .update_shell_completions! generates shell completions

Failure/Error: described_class.update_shell_completions!

NoMethodError:
  undefined method `split' for nil:NilClass

          cmd_parser.description.split(DESCRIPTION_SPLITTING_PATTERN).first
                                ^^^^^^
```

**Pull request merged**
